### PR TITLE
Type definitions

### PR DESCRIPTION
**Pull request by wkillerud**
For #153 

I think we're getting close with this. I've done some smoke tests in this repo and my own using `npm link`.

I'm having an issue in my project using `npm link`: `this.$nuxtSocket()` is suddenly no longer available (`Property '$nuxtSocket' does not exist on type 'CombinedVueInstance...`). When I check the same in this repo VS Code finds the type declarations just fine. And other things work as expected with `npm link`. I find `context.$nuxtSocket()`, the typings for the configuration work. It seems like the augmentation of Vue types fails for some reason when I use `npm link`...

I'll leave it here for today and continue debugging some other day, maybe tomorrow. I'll allow edits from maintainers, so feel free to pick up where I left off if you have the time and energy :)

### CODE REVIEW

**Comment by wkillerud** [comment]
I tried `npm install git://github.com/wkillerud/nuxt-socket-io#feat/types-dev` instead of `npm link` and my issue with `Property '$nuxtSocket' does not exist...` went away by magic :mage: Shouldn't be a problem in a published package, in other words. My test project now builds fine.

My ESLint complained with `error    NuxtSocket not found in 'nuxt-socket-io'  import/named`. May be due to benmosher/eslint-plugin-import#1699. Their suggested workaround of using `import type { NuxtSocket } from 'nuxt-socket-io';` works well.

Updated usage example:

```
import Vue from 'vue';
import type { NuxtSocket } from 'nuxt-socket-io';

export default Vue.extend({
  data() {
    return {
      message: '',
      socket: null as NuxtSocket | null,
      socketStatus: {},
    };
  },

  mounted(): void {
    this.socket = this.$nuxtSocket({});
    this.socket.on('connect', () => {
      if (!this.socket) {
        return;
      }
      console.log('🚀');
    });
  },
});
```

In other words I think this is good to go after some polish :sparkles:

File: package.json [hunk 1]
```diff
@@ -47,6 +48,8 @@
     "utils"
   ],
   "dependencies": {
+    "@nuxt/types": "^2.14.7",
+    "@types/socket.io-client": "^1.4.34",
```
Comment (richardeschloss) [inline package.json:52]: I think these should go in "devDependencies"?

**Comment by richardeschloss** [comment]
Hi! Thanks for submitting this PR! I really appreciate it and it looks like really good work so far. I'll try to get it reviewed today.

File: package.json (hunk 1 above)
Comment (richardeschloss) [reply package.json:52]: Oh wait, I think dependencies is fine.

File: io/types.d.ts [hunk 2]
```diff
@@ -1,57 +1,203 @@
...
+import Vue from 'vue';
+
+/**
+ * The format of each entry in mutations can be:
+ *
+ *   1. A single name string - the event name acts as the mutation
+ *   2. A string with a double-dashed arrow - the left side of the arrow is the event name, the right side is the mutation
+ *   3. An object - the object key is the event, the value is the mutation
+ */
+type MutationNotation = string | Record<string | string>;
```
Comment (richardeschloss) [inline io/types.d.ts:12]: I think the Record should be: Record<string, string> ?

File: io/types.d.ts [hunk 3]
```diff
@@ -1,57 +1,203 @@
...
+   * Specifies the property in [this] component that will be used
+   * to contain emit errors (see section below).
    * (referring to this.emitErrors, an object of arrays)
-   * @default 'emitErrors' 
+   * @default 'emitErrors'
    */
   emitErrorsProp?: string;
   namespaceCfg?: {
     emitters?: Array<string | object>;
     listeners?: Array<string | object>;
```
Comment (richardeschloss) [inline io/types.d.ts:85]: namespaceCfg can also have emitBacks: (I forgot to add that)
`Array<EmitBackNotation>;`

I like what you've done for the vuexOpts. I think similar records can be for these emitters and listeners.

File: io/types.d.ts [hunk 4]
```diff
@@ -1,57 +1,203 @@
...
+  ioApiProp?: string;
+  /**
+   * @default 'ioData'
+   */
+  ioDataProp?: string;
+  /**
+   * @default []
+   */
+  apiIgnoreEvts?: Array<string>;
+  serverAPI?: {
```
Comment (richardeschloss) [inline io/types.d.ts:99]: Can probably reuse this type. I refer to it as the KissAPI type, that are reused by both client and server.

File: io/types.d.ts [hunk 5]
```diff
@@ -1,57 +1,203 @@
...
-   * to contain emit errors (see section below). 
+  /**
+   * Specifies the property in [this] component that will be used
+   * to contain emit errors (see section below).
    * (referring to this.emitErrors, an object of arrays)
-   * @default 'emitErrors' 
+   * @default 'emitErrors'
    */
   emitErrorsProp?: string;
   namespaceCfg?: {
```
Comment (richardeschloss) [inline io/types.d.ts:83]: Similar to NuxtSocketVueOptions type, will probably be good to have NuxtSocketNspCfg type or similar.

File: io/types.d.ts [hunk 6]
```diff
@@ -1,57 +1,203 @@
...
+   * Options to let you sync incoming events to a Vuex store and emit events
+   * on Vuex store changes. These options will override settings from your Nuxt
+   * configuration.
+   */
+  vuex?: NuxtSocketVueOptions;
+  /**
+   * Socket.IO namespaces configuration. Supports an arrow syntax in each entry
+   * to help describe the flow (with pre/post hook designation support too).
+   */
+  namespaces?: Record<string, {
```
Comment (richardeschloss) [inline io/types.d.ts:140]: If using a namespaceCfg type as mentionned above, this can be Record<string, namespaceCfg> or similar.

File: io/types.d.ts [hunk 7]
```diff
@@ -1,57 +1,203 @@
...
+interface NuxtSocketIoRuntimeOptions {
+  /**
+   * Minimum one socket required.
+   */
+  sockets: Array<NuxtSocketRuntimeConfig>;
+}
+
+interface NuxtSocket extends SocketIOClient.Socket {};
+
+type Factory = (ioOpts: NuxtSocketOpts) => NuxtSocket;
```
Comment (richardeschloss) [inline io/types.d.ts:171]: :+1:

File: io/types.d.ts [hunk 8]
```diff
@@ -1,57 +1,203 @@
...
+
+interface NuxtSocketRuntimeConfig extends NuxtSocketConfig {
+  /**
+   * Name is required when using public/privateRuntimeConfig since the plugin
+   * uses it to merge the configurations together.
+   */
+  name: string;
+}
+
+interface NuxtSocketIoOptions {
```
Comment (richardeschloss) [inline io/types.d.ts:155]: There are more options supported: 1) warnings (for console warnings), 2) server (for io server). I can help put this in.

**Comment by richardeschloss** [comment]
Hi @wkillerud , ok I left PR comments and just pushed my edits on top of yours to [feat/types-dev](https://github.com/richardeschloss/nuxt-socket-io/tree/feat/types-dev). It should just be a fast-forward merge. Things seem to work pretty well. Do you want to try it on your end?

File: package.json (hunk 1 above)
Comment (wkillerud) [reply package.json:52]: Yes, these need to be in dependencies I think, so they get downloaded automatically along with `nuxt-socket-io`

File: io/types.d.ts (hunk 2 above)
Comment (wkillerud) [reply io/types.d.ts]: Good catch, thanks! :detective:

**Comment by wkillerud** [comment]
These changes look good. Thank you - nice work @richardeschloss

I pulled the latest changes in my test project and everything there still works fine. To be clear though I only use a small portion of all the features, so I can't vouch for _all_ of it :sweat_smile: But from what I can tell looking at the types, the plugin and module code everything seems correct.
